### DISCUSSION TITLE
functions no longer accepted as ReactNode in [@types/react]: 18.0.0

### DISCUSSION THREAD

**Discussion by Flammae** [DefinitelyTyped/DefinitelyTyped#60479] [Issues with a @types package]
disclaimer: I'm not sure if this is a deliberate change for react 18, but I really tried to find any information about this change and it's not mentioned anywhere.

Prior to `@types/react: 18.0.0` you could specify functions as a children to react elements. For example `<div>{() => "foo"}</div>`  was acceptable for the compiler. When you do this, React logs error to the console, but the function is ignored and not mounted. This behavior did not change in React 18, only the types did. 

`ReactNode` type is slightly different from previous version which makes functions not assignable to `ReactNode`. 

This also affects custom elements that have children typed as `ReactNode` which in Previous versions assumed inclusion of functions.

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment]
This is intenden since React, as you already noted, does not accept functions as children. We also don't accept BigInt for the same reason.

**Comment by Flammae** [comment]
While I understand compiler errors when it comes to BigInt, I don't entirely agree that functions should be treated the same way as BigInts when it comes to linting.

When I saw the error, my first instinct was that ReactNode type had to change for compatibility reasons to react 18, but If this is not the case, why change it now? For the longest time functions were treated as ReactNode, which spawned patterns in the community specific to this assumption. For example, a library I'm currently building calls React.Fragment's children if it happens to be a function. This is a code from that library: 
```jsx
if (
    React.isValidElement(child) &&
    isFragment(child) &&
    typeof child.props.children === 'function'
) {
    // Child's signature looks like <>{(props: any)=> React.ReactNode}</>.
    // Call the fragment's inner function with props
    return child.props.children(props);
}
```

In my opinion, children should not be checked strictly, because anything you provide as a children is not a final value that React renders. React Provides an API to manipulate children before React finally commits the changes. In my case, I use the API to call the Fragment's children function before passing it to React to render.

The fact that ReactNode was meant to include functions is also demonstrated by how React runtime chooses to warn when you try to render a function. When React detects a function was rendered by an element, the process is not stopped and an error is logged in the console saying: `Warning: Functions are not valid as a React child. This may happen if you return a Component instead of <Component /> from render. Or maybe you meant to call this function rather than return it`. Note that suggestion is to call that function. 

Compare this to what React does when it comes across a rendered object: 
![image](https://user-images.githubusercontent.com/22842333/169872194-7d2071d4-628c-4799-a28f-9fb05e810d77.png)
Error is thrown and the whole process is stopped.

Final case I have against this change is that current implementation is a bit iffy. For example, Typescript fails to lint the error with the shorthand React.Fragment syntax:
```jsx
// ts error: Type '() => void' is not assignable to type 'ReactNode'.
<React.Fragment>{() => {}}</React.Fragment>

// No error, just like in the previous versions.
<>{() => {}}</>
<>{[1,2,3].map(() => function () {})}</>
```
If this is a limitation of current architecture and not a bug it's a huge problem because, if you wrap things in `<></>`, you can basically get away with rendering stuff that causes errors Including BigInt and objects.

Also: 
```jsx
// ts error: This JSX tag's 'children' prop expects a single child of type 'ReactNode', but multiple children were provided.
<div>{()=> {}} <span /></div>
```
This error is misleading. Div does not expect a single child, it just does not like one of them to be a function.

please take a look at the [codesandbox example](https://codesandbox.io/s/agitated-minsky-t68eed) where I tried to demonstrate all the things I talked about.

**Reply by eps1lon** [reply]
> In my opinion, children should not be checked strictly, because anything you provide as a children is not a final value that React renders.

`ReactNode` is the type React handles not your component. If your component accepts functions as children you can include that:

```tsx
interface Props {
   children?: React.ReactNode | (() => string)
}
function Component(props: Props) {}
```

> The fact that ReactNode was meant to include functions

I'm not aware of this fact. Could you provide some context as to why this was "meant to be"? For context, we were specifically asked by React core members to exclude `{}` from `ReactNode` and therefore functions.

> The fact that ReactNode was meant to include functions is also demonstrated by how React runtime chooses to warn when you try to render a function. When React detects a function was rendered by an element, the process is not stopped and an error is logged in the console saying

There are plenty of warnings that do not crash which the types reject. So far we've always reflected what React would warn about, not what React would crash for.

> This error is misleading. Div does not expect a single child, it just does not like one of them to be a function.

This is not actionable by us. TypeScript errors are a recurring problem that are best raised against https://github.com/microsoft/typescript

**Comment by Flammae** [comment]
> If your component accepts functions as children you can include that.

My component accepts `React.Fragment`s as children. Optionally those fragments have functional children which are called by my component. Sort of like render props with a twist.

I understand I can't control Fragments, their implementation or types, and that they are prone to change overtime, but I did design my API around them anyway. (It's a ME problem)

> I'm not aware of this fact. Could you provide some context as to why this was "meant to be"? 

It's not a fact. I don't have any insider knowledge of what's meant to be or not. I used React runtime errors as a demonstration that React team thinks about some invalid values differently. Rendered `{}` throws, rendered `() => {}` warnes, rendered `BigInt` is ignored (for some reason). 

> So far we've always reflected what React would warn about, not what React would crash for.

React warns because invalid values are rendered to the DOM, not because invalid values are provided to `React.createElement`  (`const x = <div>{{}}</div>` wont crash until you render `x`). What I'm trying to say is that invalid values aren't invalid to the `React.createElement`, they are invalid to whatever function React uses to render the create-ed-element. And there are API-s that let you run code between `React.createElement` and the React's renderer, so type checks on `createElement` don't guarantee valid or invalid values.

My problem is that I **feel like** this new version is deviation from what React team had in mind for React. Unfortunately I can't check if this is true. 

Thanks for your time in unswerving my stuff

**Reply by Flammae** [reply]
> So far we've always reflected what React would warn about, not what React would crash for

In the versions prior to 18.0.0 there was very limited linting for invalid values in jsx. None of these warnings/crashes were detected by typescript. I think all these errors I came across come from the `ReactNode` change

**Reply by Flammae** [reply]
Would you recommend to raise an issue in the React Repository as well to get an answer on whether or not invalid jsx values should be checked by static typechekers? (except for `{}`, which React members want to be checked)

**Comment by Flammae** [comment]
So I asked React members for their opinion on the change and [gaeron](https://github.com/gaearon) thoroughly [answered my concerns](https://github.com/facebook/react/issues/24614#issuecomment-1137233313), saying they like the change. I guess that concludes this discussion. Thanks @eps1lon for helping out!
